README.md
## mia_hand_ros_pkgs (noetic) - 1.0.0-14

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Tue, 25 Jan 2022 10:09:19 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-13`
- old version: `1.0.0-13`
- new version: `1.0.0-14`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-13

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Mon, 24 Jan 2022 12:00:47 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-12`
- old version: `1.0.0-12`
- new version: `1.0.0-13`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-12

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 16:16:59 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-11`
- new version: `1.0.0-12`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-11

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 16:07:28 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-10`
- new version: `1.0.0-11`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-10

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 16:04:28 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-9`
- new version: `1.0.0-10`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-9

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 15:16:38 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-8`
- new version: `1.0.0-9`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-8

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 15:02:13 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-7`
- new version: `1.0.0-8`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-7

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 14:48:32 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-6`
- new version: `1.0.0-7`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-6

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 14:24:46 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-5`
- new version: `1.0.0-6`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-5

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 14:08:01 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-4`
- new version: `1.0.0-5`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-4

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 12:16:19 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-3`
- new version: `1.0.0-4`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-3

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 12:12:53 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-2`
- new version: `1.0.0-3`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-2

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs` on `Wed, 19 Jan 2022 12:02:30 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: https://github.com/Prensilia-srl/mia_hand_ros_pkgs-release.git
- rosdistro version: `1.0.0-1`
- old version: `1.0.0-1`
- new version: `1.0.0-2`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`


## mia_hand_ros_pkgs (noetic) - 1.0.0-1

The packages in the `mia_hand_ros_pkgs` repository were released into the `noetic` distro by running `/usr/bin/bloom-release --rosdistro noetic --track noetic mia_hand_ros_pkgs --edit` on `Wed, 01 Dec 2021 10:49:52 -0000`

These packages were released:
- `mia_hand_bringup`
- `mia_hand_description`
- `mia_hand_driver`
- `mia_hand_gazebo`
- `mia_hand_moveit_config`
- `mia_hand_msgs`
- `mia_hand_ros_control`
- `mia_hand_ros_pkgs`

Version of package(s) in repository `mia_hand_ros_pkgs`:

- upstream repository: https://bitbucket.org/prensiliasrl/mia_hand_ros_pkgs.git
- release repository: unknown
- rosdistro version: `null`
- old version: `null`
- new version: `1.0.0-1`

Versions of tools used:

- bloom version: `0.10.7`
- catkin_pkg version: `0.4.24`
- rosdep version: `0.21.0`
- rosdistro version: `0.8.3`
- vcstools version: `0.1.42`




CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mia_hand_ros_control
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2022-01-25)
------------------
* 1.0.0
* Updated CHANGELOG files.
* Updated CHANGELOG files.
* Added srv dependency to mia_hand_driver and joint_state_interface to mmia_hand_description. Added wiki url links.
* Added license file in the repo and license text as header of each file.
* Added licence file and licence text as header of each file.
* 1.0.0
* Added ChangeLog files of each pkg.
* Initial commit.
* Fix package xml files.
* Initial commit.
* Contributors: Andrea Burani, frcini


mia_hand_driver/src/serial_port.cpp
#include "mia_hand_driver/serial_port.h"

namespace mia_hand
{
FingerSerialInfo::FingerSerialInfo():
  mot_pos(0),
  mot_spe(0),
  mot_cur(0),
  fin_sg_raw{0, 0}
{

}

SerialPort::SerialPort(std::mutex* p_finger_data_mtx,
                       std::mutex* p_connection_mtx, bool* p_is_connected):
  p_finger_data_mtx_(p_finger_data_mtx),
  p_connection_mtx_(p_connection_mtx),
  p_is_connected_(p_is_connected)
{

}

SerialPort::~SerialPort()
{
  if (this->IsOpen())
  {
    this->Close();
  }
}

bool SerialPort::open(uint16_t port_num)
{
  if (this->IsOpen())
  {
    this->Close();
  }

  bool no_exceptions = true;

  std::string port_name = "/dev/ttyUSB";
  port_name += std::to_string(port_num);

  try
  {
    this->Open(port_name);
  }
  catch (LibSerial::OpenFailed& e)
  {
    no_exceptions = false;
  }
  catch (std::runtime_error& e)
  {
    no_exceptions = false;
  }

  if (no_exceptions)
  {
    this->SetBaudRate(LibSerial::BaudRate::BAUD_115200);
    this->FlushIOBuffers();
  }

  return no_exceptions;
}

bool SerialPort::close()
{
  bool no_exceptions = true;

  try
  {
    this->Close();
  }
  catch (LibSerial::NotOpen& e)
  {
    // Keep program running.
  }
  catch (std::runtime_error& e)
  {
    no_exceptions = false;
  }

  return no_exceptions;
}

void SerialPort::sendCommand(const std::string& command)
{
  serial_write_mtx_.lock();

  try
  {
    Write(command);
    WriteByte((char) 0x2A);
    WriteByte((char) 0x0D);
  }
  catch (LibSerial::NotOpen& e)
  {
    // TODO: decide if notifying somehow the exception or not.
  }
  catch (std::runtime_error& e)
  {
    // TODO: decide if notifying somehow the exception or not.
  }

  serial_write_mtx_.unlock();

  return;
}

void SerialPort::parseStream(FingerSerialInfo& thumb,
                             FingerSerialInfo& index,
                             FingerSerialInfo& mrl,
                             bool& is_checking_on)
{
  bool no_exceptions = true;

  try
  {
    this->ReadLine(stream_msg_, '\n', 1000);
  }
  catch (LibSerial::NotOpen& e)
  {
    no_exceptions = false;
  }
  catch (std::runtime_error& e)
  {
    no_exceptions = false;

    p_connection_mtx_->lock();

    if (is_checking_on)
    {
      *p_is_connected_ = false;
    }

    p_connection_mtx_->unlock();
  }

  if (no_exceptions)
  {
    switch (stream_msg_[0])
    {
      case '<':

        if ('?' == stream_msg_[1])
        {
          p_connection_mtx_->lock();
          is_checking_on = false;
          *p_is_connected_ = true;
          p_connection_mtx_->unlock();
        }

      break;

      case 'e':

        p_finger_data_mtx_->lock();

        thumb.mot_pos = (stream_msg_[9]  - 48) * 100
                      + (stream_msg_[10] - 48) * 10 + stream_msg_[11] - 48;

        if ('-' == stream_msg_[6])
        {
          thumb.mot_pos = -thumb.mot_pos;
        }

        index.mot_pos = (stream_msg_[27] - 48) * 100
                      + (stream_msg_[28] - 48) * 10 + stream_msg_[29] - 48;

        if ('-' == stream_msg_[24])
        {
          index.mot_pos = -index.mot_pos;
        }

        mrl.mot_pos   = (stream_msg_[18] - 48) * 100
                      + (stream_msg_[19] - 48) * 10 + stream_msg_[20] - 48;

        if ('-' == stream_msg_[15])
        {
          mrl.mot_pos = -mrl.mot_pos;
        }

        p_finger_data_mtx_->unlock();

      break;

      case 's':

        p_finger_data_mtx_->lock();

        thumb.mot_spe = (stream_msg_[10] - 48) * 10 + stream_msg_[11] - 48;

        if ('-' == stream_msg_[6])
        {
          thumb.mot_spe = -thumb.mot_spe;
        }

        index.mot_spe = (stream_msg_[28] - 48) * 10 + stream_msg_[29] - 48;

        if ('-' == stream_msg_[24])
        {
          index.mot_spe = -index.mot_spe;
        }

        mrl.mot_spe   = (stream_msg_[19] - 48) * 10 + stream_msg_[20] - 48;

        if ('-' == stream_msg_[15])
        {
          mrl.mot_spe = -mrl.mot_spe;
        }

        p_finger_data_mtx_->unlock();

      break;

      case 'c':

        p_finger_data_mtx_->lock();

        thumb.mot_cur = (stream_msg_[8]  - 48) * 1000
                      + (stream_msg_[9]  - 48) * 100
                      + (stream_msg_[10] - 48) * 10 + stream_msg_[11] - 48;

        if ('-' == stream_msg_[6])
        {
          thumb.mot_cur = -thumb.mot_cur;
        }

        index.mot_cur = (stream_msg_[26] - 48) * 1000
                      + (stream_msg_[27] - 48) * 100
                      + (stream_msg_[28] - 48) * 10 + stream_msg_[29] - 48;

        if ('-' == stream_msg_[24])
        {
          index.mot_cur = -index.mot_cur;
        }

        mrl.mot_cur = (stream_msg_[17] - 48) * 1000
                    + (stream_msg_[18] - 48) * 100
                    + (stream_msg_[19] - 48) * 10 + stream_msg_[20] - 48;

        if ('-' == stream_msg_[15])
        {
          mrl.mot_cur = -mrl.mot_cur;
        }

        p_finger_data_mtx_->unlock();

      break;

      case 'a':

        p_finger_data_mtx_->lock();

        thumb.fin_sg_raw[0] = (stream_msg_[35] - 48) * 1000
                            + (stream_msg_[36] - 48) * 100
                            + (stream_msg_[37] - 48) * 10
                            + stream_msg_[38] - 48;

        if ('-' == stream_msg_[33])
        {
          thumb.fin_sg_raw[0] = -thumb.fin_sg_raw[0];
        }

        thumb.fin_sg_raw[1] = (stream_msg_[44] - 48) * 1000
                            + (stream_msg_[45] - 48) * 100
                            + (stream_msg_[46] - 48) * 10
                            + stream_msg_[47] - 48;

        if ('-' == stream_msg_[42])
        {
          thumb.fin_sg_raw[1] = -thumb.fin_sg_raw[1];
        }

        index.fin_sg_raw[0] = (stream_msg_[26] - 48) * 1000
                            + (stream_msg_[27] - 48) * 100
                            + (stream_msg_[28] - 48) * 10
                            + stream_msg_[29] - 48;

        if ('-' == stream_msg_[24])
        {
          index.fin_sg_raw[0] = -index.fin_sg_raw[0];
        }

        index.fin_sg_raw[1] = (stream_msg_[17] - 48) * 1000
                            + (stream_msg_[18] - 48) * 100
                            + (stream_msg_[19] - 48) * 10
                            + stream_msg_[20] - 48;

        if ('-' == stream_msg_[15])
        {
          index.fin_sg_raw[1] = -index.fin_sg_raw[1];
        }

        mrl.fin_sg_raw[0]   = (stream_msg_[8]  - 48) * 1000
                            + (stream_msg_[9]  - 48) * 100
                            + (stream_msg_[10] - 48) * 10
                            + stream_msg_[11] - 48;

        if ('-' == stream_msg_[6])
        {
          mrl.fin_sg_raw[0] = -mrl.fin_sg_raw[0];
        }

        mrl.fin_sg_raw[1]   = (stream_msg_[53] - 48) * 1000
                            + (stream_msg_[54] - 48) * 100
                            + (stream_msg_[55] - 48) * 10
                            + stream_msg_[56] - 48;

        if ('-' == stream_msg_[51])
        {
          mrl.fin_sg_raw[1] = -mrl.fin_sg_raw[1];
        }

        p_finger_data_mtx_->unlock();

      break;

      default:

      break;
    }
  }

  return;
}
}  // namespace
